flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/VertexFinishedStateChecker.java

```diff
@@ -0,0 +1,216 @@
+ *             finished.
+ *       </ul>
+ * </ol>
+ */
+public class VertexFinishedStateChecker {
+
+    private final Set<ExecutionJobVertex> vertices;
+
+    private final Map<OperatorID, OperatorState> operatorStates;
+
+    public VertexFinishedStateChecker(
+            Set<ExecutionJobVertex> vertices, Map<OperatorID, OperatorState> operatorStates) {
+        this.vertices = vertices;
+        this.operatorStates = operatorStates;
+    }
+
+    public void validateOperatorsFinishedState() {
+        VerticesFinishedStatusCache verticesFinishedCache =
+                new VerticesFinishedStatusCache(operatorStates);
+        for (ExecutionJobVertex vertex : vertices) {
+            VertexFinishedState vertexFinishedState = verticesFinishedCache.getOrUpdate(vertex);
+
+            if (vertexFinishedState == VertexFinishedState.FULLY_FINISHED) {
+                checkProcessorsOfFullyFinishedVertex(vertex, verticesFinishedCache);
+            } else if (vertexFinishedState == VertexFinishedState.PARTLY_FINISHED) {
+                checkProcessorsOfPartlyFinishedVertex(vertex, verticesFinishedCache);
+            }
+        }
+    }
+
+    private void checkProcessorsOfFullyFinishedVertex(
+            ExecutionJobVertex vertex, VerticesFinishedStatusCache verticesFinishedStatusCache) {
+        boolean allPredecessorsFinished =
+                vertex.getInputs().stream()
+                        .map(IntermediateResult::getProducer)
+                        .allMatch(
+                                jobVertex ->
+                                        verticesFinishedStatusCache.getOrUpdate(jobVertex)
+                                                == VertexFinishedState.FULLY_FINISHED);
+
+        if (!allPredecessorsFinished) {
+            throw new FlinkRuntimeException(
+                    "Illegal JobGraph modification. Cannot run a program with fully finished"
+                            + " vertices predeceased with the ones not fully finished. Task vertex "
+                            + vertex.getName()
+                            + "("
+                            + vertex.getJobVertexId()
+                            + ")"
+                            + " has a predecessor not fully finished");
+        }
+    }
+
+    private void checkProcessorsOfPartlyFinishedVertex(
+            ExecutionJobVertex vertex, VerticesFinishedStatusCache verticesFinishedStatusCache) {
+        // Computes the distribution pattern from the predecessors. If there are multiple edges,
+        // ALL_TO_ALL edges would have a higher priority.
+        Map<JobVertexID, DistributionPattern> predecessorDistribution = new HashMap<>();
+        for (JobEdge jobEdge : vertex.getJobVertex().getInputs()) {
+            predecessorDistribution.compute(
+                    jobEdge.getSource().getProducer().getID(),
+                    (k, v) ->
+                            v == DistributionPattern.ALL_TO_ALL
```

Comment:
Is it for cases when a vertex is connected twice to the current vertex? If so, could you add such a comment here?

Reply:
I updated the comments above these lines to be more specific~